Add user service

akashmane1598 opened the pull request:
## Description
Add user service which saves user data in memory.

<!--
- **on a feature**: describe the feature and how this change fits in it, e.g. this PR makes kafka message.max.bytes configurable to better support batching
- **on a refactor**: describe why this is better than previous situation e.g. this PR changes logic for retry on healthchecks to avoid false positives
- **on a bugfix**: link relevant information about the bug (github issue or slack thread) and how this change solves it e.g. this change fixes #99999 by adding a lock on read/write to avoid data races.
-->


### Testing
Please describe the tests that you ran to verify your changes. Please summarize what did you test and what needs to be tested e.g. deployed and tested helm chart locally. 

### Checklist:
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] Any dependent changes have been merged and published in downstream modules

### Documentation
Make sure that you have documented corresponding changes in this repository or [hypertrace docs repo](https://github.com/hypertrace/hypertrace-docs-website) if required.

<!--
Include __important__ links regarding the implementation of this PR.
This usually includes and RFC or an aggregation of issues and/or individual conversations that helped put this solution together. This helps ensure there is a good aggregation of resources regarding the implementation.
-->

akashmane1598 committed:
feat: add user service

cseas commented on projects/common/src/user/user.service.ts:
```diff
@@ -0,0 +1,33 @@
+@Injectable({
+  providedIn: 'root'
+})
+export class UserService {
```
Please rename to UserInfoService or something else. The name `UserService` may easily be confused that it relates to the backend "Hypertrace User Service".

https://hus.concierge.stage.razorpay.in/swagger/index.html

akashmane1598 committed:
fix: rename to user-info

jaywalker21 commented on projects/common/src/telemetry/providers/rudderstack/rudderstack-provider.ts:
```diff
@@ -27,15 +27,8 @@ export class RudderStackTelemetry implements UserTelemetryProvider<RudderStackCo
-        console.error('something went wrong in identify method', error);
-      }
-    );
+    const data = this.userInfoService.getUserData();
```
Can do this as
`const {email, name} = this.userInfoService.getUserData()`

jaywalker21 commented on projects/common/src/user/user-info.service.ts:
```diff
@@ -0,0 +1,33 @@
+  public static readonly STORAGE_KEY: 'user-data';
+  public static readonly DEFAULT_USER: UserTraits = { name: 'Default User', email: 'ht-user@razorpay.com' };
+  public constructor(private readonly http: HttpClient, private readonly inMemoryStorage: InMemoryStorage) {}
+  public load(): void {
```
Can we return an observable here? The `APP_INITIALIZER` says that we should return a promise or observable to ensure that it is resolved.

jaywalker21 commented on projects/common/src/user/user-info.service.ts:
```diff
@@ -0,0 +1,33 @@
+export class UserInfoService {
+  public BASE_URL: string = '/user-preferences';
+  public static readonly STORAGE_KEY: 'user-data';
+  public static readonly DEFAULT_USER: UserTraits = { name: 'Default User', email: 'ht-user@razorpay.com' };
```
Let's have the `name` as `ht-user` for uniformity?

jaywalker21 commented on projects/common/src/user/user-info.service.ts:
```diff
@@ -0,0 +1,33 @@
+        // tslint:disable-next-line: no-console
+        console.error('something went wrong in while fetching user-info', error);
+      }
+    );
```
Can we do a `tap` rather than `subscribe` for the `APP_INITIALISER` part?

cseas reviewed (approved):
LGTM

jaywalker21 reviewed (commented):
Some comments, if you can go through once.

cseas commented on projects/common/src/user/user-info.service.ts:
```diff
@@ -0,0 +1,33 @@
+      },
+      error => {
+        // tslint:disable-next-line: no-console
+        console.error('something went wrong in while fetching user-info', error);
```
What happens in development mode? Do we always get this error in the console?

akashmane1598 committed:
fix: lint errs

Disable user-info api for local, object destruction for rudderstack, fix user-info module

jaywalker21 commented on projects/common/src/user/user-info.service.ts:
```diff
@@ -0,0 +1,49 @@
+  providedIn: 'root'
+})
+export class UserInfoService {
+  public BASE_URL: string = '/user-preferences';
```
nit: Not something that needs urgent update, but URL is mostly the whole link, we should have something like `USER_PREFERENCE_ENDPOINT` or something similar here.

jaywalker21 reviewed (approved):
One nit comment, else looks good
